pkg/controller/crdmirroring/crdhandler/clustergroup.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crdhandler
+
+import (
+	"context"
+	"reflect"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	crd "github.com/vmware-tanzu/antrea/pkg/apis/crd/v1alpha2"
+	crdclientset "github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned"
+	crdlister "github.com/vmware-tanzu/antrea/pkg/client/listers/crd/v1alpha2"
+	"github.com/vmware-tanzu/antrea/pkg/controller/crdmirroring/types"
+	legacycore "github.com/vmware-tanzu/antrea/pkg/legacyapis/core/v1alpha2"
+	legacycrdclientset "github.com/vmware-tanzu/antrea/pkg/legacyclient/clientset/versioned"
+	legacycorelister "github.com/vmware-tanzu/antrea/pkg/legacyclient/listers/core/v1alpha2"
+)
+
+type ClusterGroupHandler struct {
+	lister       crdlister.ClusterGroupLister
+	legacyLister legacycorelister.ClusterGroupLister
+	client       crdclientset.Interface
+	legacyClient legacycrdclientset.Interface
+}
+
+func NewClusterGroupHandler(lister crdlister.ClusterGroupLister,
+	legacyLister legacycorelister.ClusterGroupLister,
+	client crdclientset.Interface,
+	legacyClient legacycrdclientset.Interface) types.MirroringHandler {
+	mc := &ClusterGroupHandler{
+		lister:       lister,
+		legacyLister: legacyLister,
+		client:       client,
+		legacyClient: legacyClient,
+	}
+	return mc
+}
+
+// GetNewObject gets the mirroring new ClusterGroup struct
+func (c *ClusterGroupHandler) GetNewObject(namespace, name string) (metav1.Object, error) {
+	lister := c.lister
+	return lister.Get(name)
+}
+
+// AddNewObject creates the mirroring new ClusterGroup
+func (c *ClusterGroupHandler) AddNewObject(obj metav1.Object) error {
+	l := obj.(*legacycore.ClusterGroup)
+	n := c.buildNewObject(l)
+	client := c.client.CrdV1alpha2().ClusterGroups()
```

Comment:
this will construct a client everytime, why don't pass in the specific client instead when constructing the handler or getting the specific client in NewController so no extra code and construction is needed in each methods.

Reply:
Passing in the specific client is better, but this is not suit for the client with a namespace. For example, ''c.client.CrdV1alpha1().NetworkPolicies(namespace)'' @tnqn 
```
func (c *CrdV1alpha1Client) NetworkPolicies(namespace string) NetworkPolicyInterface {
	return newNetworkPolicies(c, namespace)
}

// newNetworkPolicies returns a NetworkPolicies
func newNetworkPolicies(c *CrdV1alpha1Client, namespace string) *networkPolicies {
	return &networkPolicies{
		client: c.RESTClient(),
		ns:     namespace,
	}
}
```

Reply:
Then should we handle Cluster scope CRD separately (esp. most our CRDs are cluster scoped)?

Reply:
@hongliangl maybe you can make `c.client.CrdV1alpha2().ClusterGroups()` and the corresponding legacyClient members of the struct, so that at least for cluster scope resources we don't create clients repeatedly.

Reply:
Thanks for reminding @tnqn, I missed some comments on panel `Conversation`. From the context, I got what you mean @jianjuns. I have thought of make `c.client.CrdV1alpha2().ClusterGroups()` and the corresponding legacyClient members of the struct in my mind, but I think that is not so unified then I gave up this idea.

Reply:
Modified as you suggested. @jianjuns